backend/src/main/java/com/javahelp/backend/domain/user/register/UserRegisterInteractor.java

```diff
@@ -0,0 +1,126 @@
+                && (IUserStore.getDefaultImplementation().readByUsername(boundary.getUsername()) != null)){
+            // If both username and email address exist, then return both errors
+            return new RegisterUserResponse(null, null, "Both the given email " +
+                    "address and username are already occupied. If that is previously yours, " +
+                    "consider logging in instead?");
+        }
+        if(IUserStore.getDefaultImplementation().readByEmail(boundary.getEmailAddress()) != null){
+            return new RegisterUserResponse(null, null, "The email" +
+                    " address is already occupied. If that is previously yours, consider logging " +
+                    "in instead?");
+        }
+        if(IUserStore.getDefaultImplementation().readByUsername(boundary.getUsername()) != null){
+            return new RegisterUserResponse(null, null, "The username" +
+                    " is already occupied. If that is previously yours, consider logging in instead?");
+        }
+
+        // Nothing goes wrong
+        return null;
+    }
+
+    /**
+     * A method for future different types of {@link User} that might need to validate
+     * their own defined attributes (i.e. beyond the uniqueness check in username and email address)
+     * @param boundary <T> interface extends {@link IUserRegisterInputBoundary} to retrieve input
+     *                 info
+     *
+     * @return true if assumptions for registering that type of {@link User} holds; false otherwise
+     */
+    protected boolean isValidated(T boundary){
+        /* I think this should be overridden in case to add some further checks for future user types.
+        isValidRegistering is "good enough" for clients and provider for now*/
+        return isValidRegistering(boundary) == null;
+    }
+
+    /**
+     * A method to bundle up {@link UserInfo} based on their given
+     * {@link com.javahelp.model.user.UserType}
+     */
+    protected U bundleInfo(T boundary){
+        return null; // UserInfo is designed as an abstract class so nothing can be initialized here
+    }
+
+    /**
+     * Process registering a {@link User}
+     * Note: This is only called after all registration tests are passed.
+     *
+     * @param boundary <T> interface extends {@link IUserRegisterInputBoundary} to retrieve input
+     *                 info
+     * @param user a newly generated {@link User}
+     *
+     * @return a {@link RegisterUserResponse} that contains all result info.
+     */
+    protected RegisterUserResponse registering(T boundary, User user) {
+        // Make a new user account as well as its password
+        User newUser = userStore.create(user, boundary.getPassword());
+        UserPassword password = userStore.readPassword(newUser.getStringID());
+
+        return new RegisterUserResponse(newUser, password, null);
+    }
```

Comment:
this method does not need to be protected, and should have a more specific name

Reply:
I'm not sure but I could not make them `private`. If not `protected`, then it should be `public`, but this is only reserved after checking all input for a successful case. Therefore, I'm kind of stuck there and still keep `protected`. Let me know if there is some deeper reason that I am not aware of 👍

Reply:
there is a different public entry point for this logic, I am not sure why this would need to be either public or protected. Would this ever change in subclasses or need to be accessed from outside the class?